test/cpp/jit/tests.h

```diff
@@ -1412,6 +1412,244 @@ void testDynamicDAG() {
+    fixture.checkPostCondition("q", "s", false);
+  }
+  // Move after
+  {
+    // Simple move backward
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(fixture.moveAfterTopologicallyValid("c", "a"));
+    fixture.checkPostCondition("c", "a", true);
+  }
+  {
+    // simple invalid move backward
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(!fixture.moveAfterTopologicallyValid("d", "a"));
+  }
+  {
+    // doesn't actually move anything
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(fixture.moveAfterTopologicallyValid("f", "e"));
+    fixture.checkPostCondition("f", "e", true);
+  }
+  {
+    // move backward with multiple dependencies
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(fixture.moveAfterTopologicallyValid("e", "c"));
+    fixture.checkPostCondition("e", "c", true);
+  }
+  {
+    // Move backward with non-zero working set
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(fixture.moveAfterTopologicallyValid("k", "f"));
+    fixture.checkPostCondition("k", "f", true);
+  }
+  {
+    // Simple move forward
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(fixture.moveAfterTopologicallyValid("c", "d"));
+    fixture.checkPostCondition("c", "d", true);
+  }
+  {
+    // Move forward with non-zero working set
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(fixture.moveAfterTopologicallyValid("f", "l"));
+    fixture.checkPostCondition("f", "l", true);
+  }
+
+  // Move before
+  {
+    // Simple move forward
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(fixture.moveBeforeTopologicallyValid("b", "d"));
+    fixture.checkPostCondition("b", "d", false);
+  }
+  {
+    // Simple move backward
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(fixture.moveBeforeTopologicallyValid("c", "a"));
+    fixture.checkPostCondition("c", "a", false);
+  }
+  {
+    // doesn't actually move anything
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(fixture.moveBeforeTopologicallyValid("a", "b"));
+    fixture.checkPostCondition("a", "b", false);
+  }
+  {
+    // move forward with deps
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(fixture.moveBeforeTopologicallyValid("f", "m"));
+    fixture.checkPostCondition("f", "m", false);
+  }
+  {
+    // move backward with deps
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(fixture.moveBeforeTopologicallyValid("l", "f"));
+    fixture.checkPostCondition("l", "f", false);
+  }
+
+  // check that dependencies in blocks are recognized
+  {
+    TopoMoveTestFixture fixture;
+    JIT_ASSERT(!fixture.moveAfterTopologicallyValid("l", "m"));
+    JIT_ASSERT(!fixture.moveBeforeTopologicallyValid("m", "l"));
+    JIT_ASSERT(!fixture.moveAfterTopologicallyValid("n", "l"));
+    JIT_ASSERT(!fixture.moveBeforeTopologicallyValid("l", "n"));
+  }
+
+  // Test that moveAfter(n) and moveBefore(n->next()) are not necessarily
+  // equivalent. Here, the dependency ordering is n -> o -> p.  So we can't
+  // move `n` after `o`, but we can move `n` before `p` (which pushes `o` after
```

Comment:
What happens when you call n.moveBeforeTopologicallyValid(p)?

Reply:
n -> p -> o